File: src/net/dinkla/raytracer/math/Point3D.java
package net.dinkla.raytracer.math;

import static java.lang.Math.sqrt;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 10.04.2010
 * Time: 15:17:43
 * To change this template use File | Settings | File Templates.
 */
public class Point3D extends Element3D {

    static public final Point3D ORIGIN = new Point3D(0, 0, 0);
    static public final Point3D MAX = new Point3D(Float.POSITIVE_INFINITY, Float.POSITIVE_INFINITY, Float.POSITIVE_INFINITY);
    static public final Point3D MIN = new Point3D(Float.NEGATIVE_INFINITY, Float.NEGATIVE_INFINITY, Float.NEGATIVE_INFINITY);
    static public final Point3D DEFAULT_CAMERA = new Point3D(0, 10, 10);

    public Point3D(final float x, final float y, final float z) {
        super(x, y, z);
    }

    public Point3D(final Element3D e) {
        super(e);
    }

    public Point3D plus(final Vector3D v) {
        return new Point3D(x + v.x, y + v.y, z + v.z);
    }

    public Point3D plus(final float f) {
        return new Point3D(x + f, y + f, z + f);
    }

    public Vector3D minus(final Point3D v) {
        return new Vector3D(x - v.x, y - v.y, z - v.z);
    }

    public Point3D minus(final Vector3D v) {
        return new Point3D(x - v.x, y - v.y, z - v.z);
    }

    public Point3D minus(final float f) {
        return new Point3D(x - f, y - f, z - f);
    }

}

File: src/commonMain/kotlin/net/dinkla/raytracer/math/Point3D.kt
package net.dinkla.raytracer.math

data class Point3D(
    val x: Double,
    val y: Double,
    val z: Double,
) {
    operator fun unaryMinus() = Vector3D(-x, -y, -z)

    operator fun plus(v: Vector3D) = Point3D(x + v.x, y + v.y, z + v.z)

    operator fun plus(f: Double) = Point3D(x + f, y + f, z + f)

    operator fun minus(v: Point3D): Vector3D = Vector3D(x - v.x, y - v.y, z - v.z)

    operator fun minus(v: Vector3D) = Point3D(x - v.x, y - v.y, z - v.z)

    operator fun minus(f: Double) = Point3D(x - f, y - f, z - f)

    fun sqrDistance(p: Point3D): Double {
        val dx = x - p.x
        val dy = y - p.y
        val dz = z - p.z
        return dx * dx + dy * dy + dz * dz
    }

    fun ith(axis: Axis) =
        when (axis) {
            Axis.X -> x
            Axis.Y -> y
            Axis.Z -> z
        }

    override fun toString(): String = "($x,$y,$z)"

    companion object {
        val UNIT = Point3D(1.0, 1.0, 1.0)
        val ORIGIN = Point3D(0.0, 0.0, 0.0)
        val MAX = Point3D(Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY)
        val MIN = Point3D(Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY)
        val X = Point3D(1.0, 0.0, 0.0)
        val Y = Point3D(0.0, 1.0, 0.0)
        val Z = Point3D(0.0, 0.0, 1.0)
    }
}

infix operator fun Double.times(p: Point3D) = Point3D(this * p.x, this * p.y, this * p.z)
